### DISCUSSION TITLE
How does `--dir` work with `pnpm link`?

### DISCUSSION THREAD

**Discussion by josefaidt** [pnpm/pnpm#5421] [Q&A] [answered]
Hello :wave: [from the docs](https://pnpm.io/cli/link#--dir-dir--c) I'm under the impression that I can run this command from the monorepo root to link a package from within the monorepo such as:

```bash
pnpm link --dir packages/my-package --global
```

where `my-package` has the following files:

```json
{
  "name": "my-package",
  "version": "0.1.0",
  "private": true,
  "bin": "./my-package.js"
}
```

```js
#!/usr/bin/env node
// my-package.js
console.log('hello from my package!')
```

However this does not appear to change the current working directory of where the command is executed and instead creates a symlink to the directory in `$(pnpm root -g)/node_modules`. Given the [official pnpm contributing guide](https://github.com/pnpm/pnpm/blob/main/CONTRIBUTING.md#setting-up-the-environment) also points to first changing directories with `cd` rather than using this `--dir` option I feel I am misunderstanding how this argument is expected to function.

As a side note, I've also noticed I need to run `pnpm install -g` after linking a package globally in order to install the binary in `pnpm bin -g`. Is this the expected behavior?

**Comment by zkochan** [comment]
Wasn't this fixed recently via https://github.com/pnpm/pnpm/pull/5371? cc @lvqq

**Comment by lvqq** [comment, marked as answer]
It has been fixed in v7.12.2, u need put `--dir packages/my-package` before `link`:
```
pnpm  --dir packages/my-package link --global
```

**Reply by josefaidt** [reply]
Ah I'm seeing a duplication of paths when using `--dir` before link where it looks like `process.cwd() + path.join(process.cwd() + input)`

```
pnpm --dir packages/my-package/dev link -g
```

```json
{
  "dependencies": {
    "dev": "link:/Users/josef/path/to/project-root/Users/josef/path/to/project-root/packages/my-package/dev"
  }
}
```

**Reply by josefaidt** [reply]
is there also a way to specify the dependency name? It appears to be setting the directory name instead of the name defined in the directory's `package.json` file

```json
{
  "name": "mypackage-dev",
  "version": "0.1.0",
  "private": true,
  "bin": "./mypackage-dev.js"
}
```

**Reply by lvqq** [reply]
Thanks for your feedback~

I have submitted a [PR](https://github.com/pnpm/pnpm/pull/5434) to fix it. I regard the dir option as a relative path and make a mistake to join it with cwd.

**Reply by lvqq** [reply]
And 
> specify the dependency name

Does it mean something like? 
```bash
pnpm --filter mypackage-dev link -g
```

**Reply by josefaidt** [reply]
Hey @lvqq thanks for the quick fix, I've confirmed upgrading to pnpm v7.13.0 and re-running `pnpm --dir packages/my-package/dev link -g` fixes both issues -- the link path and the package name.

Regarding the filter argument, that was coincidentally the first option I reached for and it didn't work for me (receiving ` ERROR   ERROR  Unknown option: 'recursive'`), then found the `--dir` option.

**Reply by lvqq** [reply]
Yeah,I just use it for sample to describe the feature u want. I‘m not sure if it's a bug or just not supported and maybe @zkochan could give some advice about supporting it or not

**Reply by zkochan** [reply]
That is not currently supported.

**Reply by josefaidt** [reply]
Totally fair, `--dir` makes perfect sense and reminds me of yarn's `--cwd`, `--filter` could be confusing.

**Reply by zkochan** [reply]
Well, I guess it would make sense to support it. I am just saying it was never implemented for the link command.

**Reply by lvqq** [reply]
I agree. Can I work on implementing this feature?

**Comment by lvqq** [comment]
Btw, I want to discuss the option `--dir`, it is easy to cause misunderstanding while using.

For exmple, here we have a monorepo: 
```
.
├── package.json
├── packages
│   └── packageA
│       └── package.json
└── pnpm-workspace.yaml
```
- With `pnpm --dir ./packages/packageA run <command>`, pnpm will run script defined in packageA under `./packages/packageA`
- With `pnpm run <command> --dir ./packages/packageA`, pnpm will run script defined in root directory under `./packages/packageA`


So should we add more information to the `--dir` [documentation](https://pnpm.io/pnpm-cli) about recommanding user to add option before command? Or make the behavior of `--dir` consistent whether it's before or after the command? cc @zkochan

**Reply by zkochan** [reply]
any options that are listed after the `run <cmd>` are passed to the script. This is documented here: https://pnpm.io/cli/run#options

**Reply by lvqq** [reply]
I just use run <cmd> as a example, I actually mean using the common option `--dir` before or after command is not consistent.  So is it necessary to emphasize that `--dir` should be before the command? The document here doesn't mention it: https://pnpm.io/pnpm-cli#-c-path---dir-path
